lib/dbservice_web/controllers/user_controller.ex

```diff
@@ -178,61 +175,69 @@ defmodule DbserviceWeb.UserController do
   end
 
   def get_user_sessions(conn, %{"user_id" => user_id, "quiz" => quiz_flag}) do
-    group_users = GroupUsers.get_group_user_by_user_id(user_id)
-    sessions = Enum.flat_map(group_users, &fetch_group_user_sessions(&1, quiz_flag))
+    sessions = fetch_user_sessions(user_id, quiz_flag)
     render(conn, :user_sessions, session: sessions)
   end
 
   def get_user_sessions(conn, %{"user_id" => user_id}) do
     get_user_sessions(conn, %{"user_id" => user_id, "quiz" => false})
   end
 
-  defp fetch_group_user_sessions(group_user, quiz_flag) do
-    group_id = group_user.group_id
+  # Single optimized query
+  defp fetch_user_sessions(user_id, quiz_flag) do
+    user_batch_data = get_user_batch_data(user_id)
 
-    case Groups.get_group_by_group_id_and_type(group_id, "batch") do
-      nil -> []
-      group -> fetch_group_sessions(group, quiz_flag)
+    if Enum.empty?(user_batch_data) do
+      []
+    else
+      if quiz_flag do
+        get_quiz_sessions(user_batch_data)
+      else
+        get_regular_sessions(user_batch_data)
+      end
     end
   end
 
-  defp fetch_group_sessions(group, quiz_flag) do
-    child_id = group.child_id
-    batch = Batches.get_batch!(child_id)
-    class_batch_id = batch.batch_id
-    quiz_id = batch.parent_id
-    quiz_group = Groups.get_group_by_child_id_and_type(quiz_id, "batch")
-    quiz_group_id = quiz_group.id
-
-    group_id = if quiz_flag, do: quiz_group_id, else: group.id
-    group_sessions = GroupSessions.get_group_session_by_group_id(group_id)
-
-    filter_sessions(group_sessions, quiz_flag, class_batch_id)
-    |> Enum.map(&get_session_by_id/1)
-  end
-
-  defp filter_sessions(group_sessions, quiz_flag, class_batch_id) do
-    Enum.filter(group_sessions, &should_include_session?(&1, quiz_flag, class_batch_id))
-  end
-
-  defp should_include_session?(group_session, quiz_flag, class_batch_id) do
-    case get_session_by_id(group_session) do
-      nil ->
-        false
-
-      session ->
-        session.is_active &&
-          if quiz_flag && session.meta_data["batch_id"] != "" do
-            # Split the comma-separated batch_id and check if class_batch_id is in the list
-            batch_ids = String.split(session.meta_data["batch_id"], ",")
-            class_batch_id in batch_ids and session.platform == "quiz"
-          else
-            true
-          end
+  # Get user's groups with batch information
+  defp get_user_batch_data(user_id) do
+    from(gu in GroupUser,
+      where: gu.user_id == ^user_id,
+      join: class_group in Group,
+      on: class_group.id == gu.group_id and class_group.type == "batch",
+      join: class_batch in Batch,
+      on: class_batch.id == class_group.child_id,
+      left_join: quiz_group in Group,
+      on: quiz_group.child_id == class_batch.parent_id and quiz_group.type == "batch",
```

Comment:
Why are you calling it quiz group, but it's of type batch? Maybe call it quiz batch only, no to be consistent?

Reply:
I kept the alias as quiz_group because the join is on the Group table, not the Batch table. Even though it’s a group of type "batch", naming it quiz_batch could be misleading since the source table is still Group